src/asr/asr_mc_chainer.py

```diff
@@ -0,0 +1,516 @@
+#!/usr/bin/env python
+
+# Copyright 2017 Johns Hopkins University (Shinji Watanabe)
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+
+import collections
+import copy
+import json
+import logging
+import math
+import os
+import pickle
+import six
+import sys
+
+# chainer related
+import chainer
+from chainer import cuda
+from chainer import function
+from chainer import reporter as reporter_module
+from chainer import training
+from chainer.training import extensions
+
+# espnet related
+from asr_utils import adadelta_eps_decay
+from asr_utils import CompareValueTrigger
+from asr_utils import delete_feat
+from asr_utils import make_batchset
+from asr_utils import restore_snapshot
+from e2e_asr_attctc import E2E
+from e2e_asr_attctc import Loss
+
+# for me2e
+from beamformer import NB_MVDR
+from me2e_asr_attctc import ME2E
+
+# for kaldi io
+import kaldi_io_py
+import lazy_io
+
+# rnnlm
+import lm_train
+
+# numpy related
+import matplotlib
+import numpy as np
+matplotlib.use('Agg')
+
+
+class ChainerSeqEvaluaterKaldi(extensions.Evaluator):
+    '''Custom evaluater with Kaldi reader for chainer'''
+
+    def __init__(self, iterator, target, reader, device, converter):
+        super(ChainerSeqEvaluaterKaldi, self).__init__(
+            iterator, target, device=device)
+        self.reader = reader
+        self.converter = converter
+
+    # The core part of the update routine can be customized by overriding.
+    def evaluate(self):
+        '''evaluate over iterator'''
+        iterator = self._iterators['main']
+        eval_func = self.eval_func or self._targets['main']
+
+        if self.eval_hook:
+            self.eval_hook(self)
+
+        if hasattr(iterator, 'reset'):
+            iterator.reset()
+            it = iterator
+        else:
+            it = copy.copy(iterator)
+
+        summary = reporter_module.DictSummary()
+
+        for batch in it:
+            observation = {}
+            with reporter_module.report_scope(observation):
+                # read scp files
+                # x: original json with loaded features
+                #    will be converted to chainer variable later
+                # batch only has one minibatch utterance, which is specified by batch[0]
+                x = self.converter(batch[0], self.reader)
+                with function.no_backprop_mode():
+                    eval_func(x)
+                    delete_feat(x)
+
+            summary.add(observation)
+
+        return summary.compute_mean()
+
+
+class ChainerSeqUpdaterKaldi(training.StandardUpdater):
+    '''Custom updater with Kaldi reader for chainer'''
+
+    def __init__(self, train_iter, optimizer, reader, device, converter):
```

Comment:
instead of this redefinition, how about adding a new last argument `converter=converter_kaldi`  in `asr_chainer.ChainerSeqUpdaterKaldi`

Reply:
That's right.
In my implementation, in order not to change the original asr_chainer, I just redefined it.
If you can add a new argument "converter=converter_kaldi" in the original Updater, we can just import and use it.

Reply:
In addition, we have to add "if statement" in _sum_sqnorm function.
Because the multichannel training sometimes passes the beamformer module, the gradient become None.
Therefore, we have to ignore such None gradient.

Is such modification OK for the original asr_chainer?

Reply:
I think those modifications are OK.

Reply:
I see.
Can I modify it by myself?
Or, should I ask someone to modify it?